File: Average/Java/joshimoo/Average_test.java
package average;

import org.junit.Test;
import java.util.stream.DoubleStream;
import static org.junit.Assert.*;

/**
 * @author Joshua Moody (joshimoo@hotmail.de)
 */
public class AverageTest {
    private static final double DELTA = 0.01;

    @Test
    public void testAverageInequality() throws Exception {
        double[] numbers = new double[] { 1, 2, 3, 4, 5 };

        /**
         * Mean Inequalities:
         * Some means are in a constant relationship to one another.
         * If we denote the arithmentic mean of x and y by A,
         * their geometric mean by G,
         * their harmonic mean by H,
         * their root mean square (quadratic mean) by R,
         * and their contraharmonic mean by C,
         *
         * then the following chain of inequalities is always true
         * C >= R >= A >= G >= H
         */

        double contraHarmonic = Average.contraHarmonicMean(numbers);
        double quadratic = Average.quadraticMean(numbers);
        double arithmetic = Average.arithmeticMean(numbers);
        double geometric = Average.geometricMean(numbers);
        double harmonic = Average.harmonicMean(numbers);
        assertTrue("Your Average inequalities are not correct", (contraHarmonic >= quadratic) && (quadratic >= arithmetic) && (arithmetic >= geometric) && (geometric >= harmonic));
    }

    @Test
    public void testArithmeticMean() throws Exception {
        double[] numbers = new double[] { 1, 2, 3, 4, 5 };
        double expected = DoubleStream.of(numbers).average().getAsDouble();
        assertEquals(expected, Average.arithmeticMean(numbers), DELTA);
    }

    @Test
    public void testWeightedMean() throws Exception {
        double[] numbers = new double[] { 1, 2, 3, 4, 5 };
        double[] weights = new double[] { 1, 1, 1, 1, 1 };
        double expected = DoubleStream.of(numbers).average().getAsDouble();
        assertEquals(expected, Average.weightedMean(numbers, weights), DELTA);
    }

    @Test
    public void testHarmonicMean() throws Exception {
        // n/(1/x1 + 1/x2 + ... + 1/xn)
        double[] numbers = new double[] { 1, 2, 3, 4, 5 };
        double expected = numbers.length / DoubleStream.of(numbers).map(x -> 1.0 / x).sum();
        assertEquals(expected, Average.harmonicMean(numbers), DELTA);
    }

    @Test
    public void testContraHarmonicMean() throws Exception {
        double[] numbers = new double[] { 1, 2, 3, 4, 5 };
        double expected = DoubleStream.of(numbers).map(x -> Math.pow(x, 2.0)).sum() / DoubleStream.of(numbers).sum();
        assertEquals(expected, Average.contraHarmonicMean(numbers), DELTA);
    }

    @Test
    public void testGeometricMean() throws Exception {
        // (x1*x2*...xn) ^ (1/n)
        double[] numbers = new double[] { 1, 2, 3, 4, 5 };
        double expected = Math.pow((numbers[0] * numbers[1] * numbers[2] * numbers[3] * numbers[4]), 1.0 / numbers.length);
        assertEquals(expected, Average.geometricMean(numbers), DELTA);
    }

    @Test
    public void testQuadraticMean() throws Exception {
        // sqrt((x1)^2+(x2)^2+(x3)^2+...+(xn)^2 /n)
        double[] numbers = new double[] { 1, 2, 3, 4, 5 };
        double expected = Math.sqrt(DoubleStream.of(numbers).map(x -> Math.pow(x, 2)).sum() / numbers.length);
        assertEquals(expected, Average.quadraticMean(numbers), DELTA);
    }

    @Test
    public void testGeneralizedMean() throws Exception {
        // y-root((x1)^y+(x2)^y+(x3)^y+...+(xn)^y / n)
        double[] numbers = new double[] { 1, 2, 3, 4, 5 };
        double power = 4;
        double expected = Math.pow((DoubleStream.of(numbers).map(x -> Math.pow(x, 4)).sum() / numbers.length), 1.0 / power);
        assertEquals(expected, Average.generalizedMean(numbers, power), DELTA);
    }

    @Test
    public void testMidrange() throws Exception {
        double[] numbers = new double[] { 1, 2, 3, 4, 5 };
        double expected = (numbers[0] + numbers[numbers.length - 1]) / 2.0;
        assertEquals(expected, Average.midrange(numbers), DELTA);
    }

    @Test
    public void testMedianEven() throws Exception {
        double[] numbers = new double[] { 1, 2, 3, 4, 5, 6};
        double expected = (numbers[(numbers.length / 2) - 1] + numbers[numbers.length / 2]) / 2.0;
        assertEquals(expected, Average.median(numbers), DELTA);
    }

    @Test
    public void testMedianOdd() throws Exception {
        double[] numbers = new double[] { 1, 2, 3, 4, 5 };
        double expected = numbers[((int) Math.floor(numbers.length / 2))];
        assertEquals(expected, Average.median(numbers), DELTA);
    }

    @Test
    public void testMode() throws Exception {
        // frequency based average
        double[] numbers = new double[] { 1, 2, 3, 3, 3, 4, 5 };
        double expected = 3;
        assertEquals(expected, Average.mode(numbers), DELTA);
    }
}

File: Average/Go/roman_kachanovsky/average_test.go
package roman_kachanovsky

import "testing"

func TestArithmeticMean(t *testing.T) {
	cases := []struct {
		in []int
		want float64
	}{
		{[]int{1, 2, 3, 4, 5}, 3.0},
		{[]int{1, 2, 3, 4}, 2.5},
		{[]int{5, 5, 5, 5, 5, 5, 5}, 5.0},
	}

	for _, c := range cases {
		got := ArithmeticMean(c.in)
		if got != c.want {
			t.Errorf("ArithmeticMean(%v) == %v, want %v", c.in, got, c.want)
		}
	}
}

func TestGeometricMean(t *testing.T) {
	cases := []struct {
		in []int
		want float64
	}{
		{[]int{1, 2, 3, 4, 5}, 2.605171084697352},
		{[]int{1, 2, 3, 4}, 2.2133638394006434},
		{[]int{5, 5, 5, 5, 5, 5, 5}, 5.0},
	}

	for _, c := range cases {
		got := GeometricMean(c.in)
		if got != c.want {
			t.Errorf("GeometricMean(%v) == %v, want %v", c.in, got, c.want)
		}
	}
}


func TestHarmonicMean(t *testing.T) {
	cases := []struct {
		in []int
		want float64
	}{
		{[]int{1, 2, 3, 4, 5}, 2.18978102189781},
		{[]int{1, 2, 3, 4}, 1.9200000000000004},
		{[]int{5, 5, 5, 5, 5, 5, 5}, 5.0},
	}

	for _, c := range cases {
		got := HarmonicMean(c.in)
		if got != c.want {
			t.Errorf("HarmonicMean(%v) == %v, want %v", c.in, got, c.want)
		}
	}
}

func TestQuadraticMean(t *testing.T) {
	cases := []struct {
		in []int
		want float64
	}{
		{[]int{1, 2, 3, 4, 5}, 3.3166247903554},
		{[]int{1, 2, 3, 4}, 2.7386127875258306},
		{[]int{5, 5, 5, 5, 5, 5, 5}, 5.0},
	}

	for _, c := range cases {
		got := QuadraticMean(c.in)
		if got != c.want {
			t.Errorf("QuadraticMean(%v) == %v, want %v", c.in, got, c.want)
		}
	}
}

func TestGeneralizedMean(t *testing.T) {
	cases := []struct {
		in []int
		power float64
		want float64
	}{
		{[]int{1, 2, 3, 4, 5}, 1.0, 3.0},
		{[]int{1, 2, 3, 4}, 1.0, 2.5},
		{[]int{5, 5, 5, 5, 5, 5, 5}, 1.0, 5.0},
		{[]int{1, 2, 3, 4, 5}, 1.5, 3.1688613641003776},
		{[]int{1, 2, 3, 4, 5}, 2.0, 3.3166247903554},
		{[]int{1, 2, 3, 4, 5}, 0.5, 2.8105398233187415},
	}

	for _, c := range cases {
		got := GeneralizedMean(c.in, c.power)
		if got != c.want {
			t.Errorf("GeneralizedMean(%v, %v) == %v, want %v", c.in, c.power, got, c.want)
		}
	}
}

func TestWeightedMean(t *testing.T) {
	cases := []struct {
		in []int
		weights []float64
		want float64
	}{
		{[]int{1, 2, 3, 4, 5}, []float64{0.1, 0.2, 0.2, 0.3, 0.2}, 3.3},
		{[]int{1, 2, 3, 4, 5}, []float64{0.5, 0.2, 0.1, 0.3, 0.2}, 2.615384615384616},
		{[]int{1, 2, 3, 4}, []float64{0.1, 0.2, 0.2, 0.3}, 2.8749999999999996},
		{[]int{5, 5, 5, 5, 5, 5, 5}, []float64{0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1}, 5.0},
	}

	for _, c := range cases {
		got := WeightedMean(c.in, c.weights)
		if got != c.want {
			t.Errorf("WeightedMean(%v, %v) == %v, want %v", c.in, c.weights, got, c.want)
		}
	}
}

func TestMidrangeMean(t *testing.T) {
	cases := []struct {
		in []int
		want float64
	}{
		{[]int{1, 2, 3, 4, 5}, 3.0},
		{[]int{1, 2, 3, 4}, 2.5},
		{[]int{5, 5, 5, 5, 5, 5, 5}, 5.0},
	}

	for _, c := range cases {
		got := MidrangeMean(c.in)
		if got != c.want {
			t.Errorf("MidrangeMean(%v) == %v, want %v", c.in, got, c.want)
		}
	}
}

func TestEnergaticMean(t *testing.T) {
	cases := []struct {
		in []int
		want float64
	}{
		{[]int{1, 2, 3, 4, 5}, 3.2276678141732704},
		{[]int{1, 2, 3, 4}, 2.6428450251024453},
		{[]int{5, 5, 5, 5, 5, 5, 5}, 4.999999999999999},
	}

	for _, c := range cases {
		got := EnergaticMean(c.in)
		if got != c.want {
			t.Errorf("EnergaticMean(%v) == %v, want %v", c.in, got, c.want)
		}
	}
}